feat: implement contactor fault

seixasxbr opened the pull request:
# What is this PR for
Simulate the contactor fault
Refactor the fault state update 
implement the hard and the soft reset

# How I did it
There is a probability associated to the emission of the contactor fault at each update hook and a probability associate to the success of the soft reset.
The hard reset  will kill the component and will always reset the fault (since the fault state is none at the start hook)

# Checklist
- [x] Assign yourself  in the PR
- [x] Write unit tests (when relevant)

seixasxbr committed:
feat: simulate the contactor fault emission

seixasxbr committed:
chore: rename gpio

the gpio enables the propulsion, so a high value indicates that the propulsion should be enabled, not that the estop is active

seixasxbr committed:
fix: remove hard reset probability

the hard reset will kill the component and will always reset the fault state

seixasxbr committed:
feat: implement soft reset attempt and refactor fault state update

wvmcastro reviewed (commented):
I think that you got it wrong: It is not the controller component that attempts the soft reset once it is in a contactor fault. It is the `ContactorFaultResetTask` that does it based on the controller's component `fault_state_port`

seixasxbr commented:
> I think that you got it wrong: It is not the controller component that attempts the soft reset once it is in a contactor fault. It is the `ContactorFaultResetTask` that does it based on the controller's component `fault_state_port`

I know that. This component won't do the soft reset, It will simulate the success of the soft reset attempt, as you can see in the code. Maybe a name chance you make this clearer

seixasxbr committed:
feat: add hard reset simulation

wvmcastro commented on tasks/SimulationTask.cpp:
```diff
@@ -108,19 +107,10 @@ InverterStatus SimulationTask::inverterStatus() const
     return InverterStatus::STATUS_READY;
 }
 
-std::uint16_t SimulationTask::currentFault() const
```
I think it would be clearer to manage all fault state here, something like:
```cpp
void SimulationTask::computeCurentFault()
{
  if(m_current_fault == Fault::CONTACTOR_FAULT) {
    if(m_external_fault && exitContactorFault()) {
      m_current_fault = Fault::EXTERNAL_FAULT; 
    }
    return m_current_fault;
  }

  if (contactorFault()) {
    return m_current_fault = Fault::CONTACTOR_FAULT;
  }
  if (m_external_fault) {
    return m_current_fault = Fault::EXTERNAL_FAULT;
  }
  
  return m_current_fault = Fault::NO_FAULT;
}
```

where `contactorFault()` and `extiContactorFault()` would use the probabilities. The update hook would be something as 
```cpp
void SimulationTask::updateHook() 
{
  ...
  readExternalFaultGPIOState();
  computeCurrentFault();
  ...
  evaluateInverterStatus(...
}
```

seixasxbr committed:
fix: contactor fault trigger

seixasxbr committed:
chore: rename method to make it more flexible

seixasxbr committed:
test: contactor fault emission, contactor fault reset and hard reset

seixasxbr committed:
chore: make the distribution generator a class member

seixasxbr committed:
chore: rename variable and improve documentation

seixasxbr committed:
chore: centralize fault state update

seixasxbr committed:
chore: remove hard reset abstraction

seixasxbr committed:
fix: only one power gpio input

since this simulates only one motor

seixasxbr committed:
fix: make the distributions class members

seixasxbr committed:
feat: read power disable gpio before

seixasxbr committed:
fix: fault state method logic

seixasxbr committed:
fix: publish fault at error hook after the update of the fault state

the publish was being done before the update, to publish the fault just after the error emission. I moved this publish to the update hook and fixed the ordering at the error hook

seixasxbr committed:
fix: do fault state port assertions

jhonasiv commented on tasks/SimulationTask.cpp:
```diff
@@ -195,9 +197,57 @@ control_base::SaturationSignal saturationSignal(bool saturation)
+
+void SimulationTask::updateFaultState()
+{
+    if (m_current_fault_state == Fault::CONTACTOR_FAULT && m_external_fault &&
+        exitContactorFault()) {
+        m_current_fault_state = Fault::EXTERNAL_FAULT;
+        return;
+    }
```
Ah, and making "state evaluation" methods that DONT assign to a member variable is usually a very good practice when you take into account unit testing.

jhonasiv replied:
https://docs.google.com/document/d/16zBGRa58zu7wgRBcGb6IVVQtCfiDqDDhu4vjUVnVSu0/edit?tab=t.0#heading=h.mq8v0to2d5n

wvmcastro commented on tasks/SimulationTask.cpp:
```diff
@@ -154,15 +143,28 @@ void SimulationTask::updateHook()
-
     readExternalFaultGPIOState();
+    updateFaultState();
+    readPortPowerDisableGPIOState();
```
@seixasxbr I think this problem still remains. Also, note that the controller can be turned off while on `errorHook`

seixasxbr commented on tasks/SimulationTask.cpp:
```diff
@@ -195,9 +197,57 @@ control_base::SaturationSignal saturationSignal(bool saturation)
+        return;
+    }
+    if (!m_external_fault && m_current_fault_state == Fault::EXTERNAL_FAULT &&
+        m_current_fault_state != Fault::CONTACTOR_FAULT) {
```
this was outdated

seixasxbr committed:
chore: use functional style

seixasxbr committed:
fix: read power disable gpio port at error hook

seixasxbr committed:
test: behaviour after io_timeout

seixasxbr committed:
chore: improve documentation

wvmcastro commented on tasks/SimulationTask.cpp:
```diff
@@ -195,16 +210,53 @@ control_base::SaturationSignal saturationSignal(bool saturation)
+        }
+        return Fault::CONTACTOR_FAULT;
+    }
+    if (!external_fault && current_fault_state == Fault::EXTERNAL_FAULT) {
+        return Fault::NO_FAULT;
+    }
+    if (triggerContactorFault()) {
+        return Fault::CONTACTOR_FAULT;
+    }
```
In "real life" I think the contactor fault has precedence over any other fault, but I do see that this order favors component testing. I think a comment clarifying that would be welcome

wvmcastro commented on test/simulation_task_test.rb:
I am missing the test for checking the component can emit an IO_TIMEOUT while on error hook

wvmcastro reviewed (approved):
write the comment and the test and this is good to go

seixasxbr committed:
chore: document fault state update

seixasxbr committed:
test: fix io_timeout emission on error hook
